core/store/src/trie/outgoing_metadata.rs

```diff
@@ -0,0 +1,451 @@
+    V0(ReceiptGroupV0),
+}
+
+#[derive(Debug, BorshSerialize, BorshDeserialize, ProtocolSchema)]
+pub struct ReceiptGroupV0 {
+    /// Total size of receipts in this group.
+    /// Should be no larger than `max_receipt_size`, otherwise the bandwidth
+    /// scheduler will not be able to grant the bandwidth needed to send
+    /// the receipts in this group.
+    pub size: u64,
+    /// Total gas of receipts in this group.
+    pub gas: u128,
+}
+
+impl ReceiptGroup {
+    /// Create a new group which will contain one receipt with this size and gas.
+    pub fn new(receipt_size: ByteSize, receipt_gas: Gas) -> ReceiptGroup {
+        ReceiptGroup::V0(ReceiptGroupV0 { size: receipt_size.as_u64(), gas: receipt_gas.into() })
+    }
+
+    /// Total size of receipts in this group.
+    pub fn size(&self) -> u64 {
+        match self {
+            ReceiptGroup::V0(group) => group.size,
+        }
+    }
+
+    pub fn size_mut(&mut self) -> &mut u64 {
+        match self {
+            ReceiptGroup::V0(group) => &mut group.size,
+        }
+    }
+
+    /// Total gas of receipts in this group.
+    pub fn gas(&self) -> u128 {
+        match self {
+            ReceiptGroup::V0(group) => group.gas,
+        }
+    }
+
+    pub fn gas_mut(&mut self) -> &mut u128 {
+        match self {
+            ReceiptGroup::V0(group) => &mut group.gas,
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        // There are no receipts with size equal to zero.
+        // Total size being zero means that there are no receipts in this group.
+        let empty = self.size() == 0;
+        if empty {
+            assert_eq!(self.gas(), 0, "Gas should be zero for an empty group.");
+        }
+        empty
+    }
+}
+
+/// Parameters which control size and gas of receipt groups stored in the ReceiptGroupsQueue.
+/// There is a lower bound and upper bound on size and gas.
+/// By default receipts are added to the last group in the queue.
+/// A new group is started if:
+///     A) adding a new receipt would make the group exceed the upper bound
+///     B) the last group already has size or gas above the lower bound
+///
+/// This way we ensure that the size and gas of the groups are within the bounds,
+/// as long as size/gas of a single receipt doesn't exceed the upper bound.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct ReceiptGroupsConfig {
+    /// All receipt groups aim to have a size above this threshold.
+    pub size_lower_bound: ByteSize,
+    /// All receipt groups aim to have a size below this threshold.
+    /// Should be no larger than `max_receipt_size`, otherwise the bandwidth
+    /// scheduler will not be able to grant the bandwidth needed to send
+    /// the receipts in this group.
+    pub size_upper_bound: ByteSize,
+    /// All receipt groups aim to have gas above this threshold.
+    pub gas_lower_bound: Gas,
+    /// All receipt groups aim to have gas below this threshold.
+    pub gas_upper_bound: Gas,
+}
```

Comment:
I spent some more time thinking about the optimal lower and upper bound sizes for receipt groups, and I think I'll:
* Remove the lower bound
* Reduce the upper bound to something like 100kB

This means that a new group will be started if adding a new receipt to the group would cause its size to go above 100kB.

As a result all groups will be below 100kB, except for the groups that contain a single receipt with size above 100kB.

AFAIU having groups like this will produce optimal bandwidth requests - identical to the ones that we would produce if we walked over all of receipts individually.

The diff between two values that can be requested is ~110kB. If all groups were below 100kB we would always produce optimal requests - each consecutive group would either request a value or move to the next value, just like individual receipts.

When there are receipts larger than 100kB we will naturally skip some of the values, but this is identical to what would happen if we operated on individual receipts. The next group smaller than 100kB will either request the same value or move to the next one, just like individual receipts.

A bit handwavy, but I feel that it works. I can make some drawings if needed.

Reply:
Yeah, makes sense, I honestly had the same thought for why were we maintaining both lower and upper bounds and we could instead do with just one lower or upper bound, however I didn't raise it in the original review.

One thing that wasn't clear to me was something on the lines of what happens when we have the following scenarios?
- 99 number of 1kb receipts
- 1 number of 500 kb receipt.

Should the ideal/expected behavior be case A or case B?

Case A
Group 1: 99 receipts with 99kb
Group 2: 1 receipt with 500 kb

Case B
Group 1: 100 receipts with 599 kb

I would imagine case A is probably more ideal than case B for bandwidth scheduling but it's harder to "code it up" compared to a simple upper bound impl?

Reply:
Which is why I had originally not questioned the upper and lower bound

Reply:
My initial design was (B). The main reason was that I wanted to have the guarantee that every group is at least 90kB large to be able to guarantee that the size of the list of groups is small enough. This was important when all receipt groups were kept in one trie value. Now that doesn't really matter.

(B) keeps the number of groups lower, but it has serious disadvantages.
The first disadvantage is that a group can get larger than `max_receipt_size`, and there's no guarantee that the bandwidth scheduler will be able to grant this much bandwidth. To deal with this I initially read the first receipt and subtracted its size from the first group, but this causes witness size issues because the first receipt could be big. To deal with this I added the `size_upper_bound`  to keep size of receipt groups below `max_receipt_size`.
The second disadvantage is that it doesn't produce optimal bandwidth requests. Smaller groups would be more precise.

In the end I moved to (A) because it doesn't have any of these disadvantages. The number of groups can be a bit larger, but that shouldn't matter when it's a TrieQueue.

Reply:
Sounds good! :)